### PR TITLE
Change buy/sell strategy

### PR DESCRIPTION
**Pull request by ortschun**
buy strategy makes the decision of buying 1% less of the current price
This kicks off sell orders that checks the buy-order status, and once that's filled, it places a sell order

Also brought a flag for specifying maximum buy orders available at a given time, to max out the risk of buying/and for testing.
Put a flag to input existing order ids. In case things go bad and we restart, we still want to keep working on selling the currently open buys.

### CODE REVIEW

File: coinmx/build.gradle [hunk 1]
```diff
@@ -18,5 +18,7 @@ dependencies {
 	compile group: 'org.knowm.xchange', name: 'xchange-binance', version: '4.3.1'
     compile group: 'com.google.guava', name: 'guava', version: '23.5-jre'
 	compile group: 'org.reflections', name: 'reflections', version: '0.9.11'
+	compile group: 'org.mapdb', name: 'mapdb', version: '3.0.5'
```
Comment (ortschun) [inline coinmx/build.gradle]: will be used later for persistent buy-order ids

File: coinmx/src/main/java/com/coinmx/config/StrategyExecutorConfiguration.java [hunk 2]
```diff
@@ -2,16 +2,34 @@
...
+    private final int maxBuysAtOnce;
+    private final List<String> existingOrderIds;
 
     @JsonCreator
-    public StrategyExecutorConfiguration(@JsonProperty("currency") String currencyName) {
+    public StrategyExecutorConfiguration(@JsonProperty("currency") String currencyName,
+                                         @JsonProperty("maxBuysAtOnce") Integer maxBuysAtOnce,
+                                         @JsonProperty("existingOrderIds") List<String> existingOrderIds) {
         this.currencyName = currencyName;
+        this.maxBuysAtOnce = MoreObjects.firstNonNull(maxBuysAtOnce, 1);
```
Comment (ortschun) [inline coinmx/src/main/java/com/coinmx/config/StrategyExecutorConfiguration.java]: defaults to 1 to play safe

File: coinmx/src/main/java/com/coinmx/strategy/StrategyExecutor.java [hunk 3]
```diff
@@ -52,22 +80,102 @@ public static synchronized StrategyExecutor getInstance(
...
+
+                BigDecimal buyPrice = ticker.getAsk().subtract(ticker.getAsk().multiply(BUY_MARGIN));
+                // HACKHACK: bigdecimal hangs in division so we need this hack
+                // also they don't accept non-integer amount to buy
+                BigDecimal amount = maxSpend.divideToIntegralValue(buyPrice);
+                buyPrice = buyPrice.setScale(NUM_DECIMAL_DIGITS, RoundingMode.FLOOR);
+
+                Optional<String> orderIdOptional = orderManager.buyLimitOrder(currency, buyPrice, amount);
+                if (orderIdOptional == null) {
+                    LOG.error("Error placing order.");
```
Comment (ogzd) [inline coinmx/src/main/java/com/coinmx/strategy/StrategyExecutor.java]: return

File: coinmx/src/main/java/com/coinmx/strategy/StrategyExecutor.java [hunk 4]
```diff
@@ -52,22 +80,102 @@ public static synchronized StrategyExecutor getInstance(
...
+        return () -> {
+            while (true) {
+                try {
+                    BinanceOrder order = orderManager.getOrderStatus(currency, orderId);
+                    BigDecimal sellPrice = order.price.add(order.price.multiply(BENEFIT_MARGIN));
+                    if (order.status == OrderStatus.FILLED) {
+                        sellPrice = sellPrice.setScale(NUM_DECIMAL_DIGITS, RoundingMode.CEILING);
+                        Optional<String> id = orderManager.sellLimitOrder(currency, sellPrice, order.executedQty);
+                        if (id.isPresent()) {
+                            LOG.info("Put a sell order {} for price: {} and amount: {}", orderId, sellPrice, order.executedQty);
```
Comment (ogzd) [inline coinmx/src/main/java/com/coinmx/strategy/StrategyExecutor.java]: if not present?

File: coinmx/src/main/java/com/coinmx/strategy/StrategyExecutor.java [hunk 5]
```diff
@@ -52,22 +80,102 @@ public static synchronized StrategyExecutor getInstance(
...
+                        }
+                        numCurrentOpenBuys.decrementAndGet();
+                        break;
+                    }
+                    else {
```
Comment (ogzd) [inline coinmx/src/main/java/com/coinmx/strategy/StrategyExecutor.java]: what if we also check if the buy order is partially filled >75%, cancel the rest of the buy order and place a sell order for the ones that are bought

File: coinmx/src/main/java/com/coinmx/strategy/StrategyExecutor.java [hunk 6]
```diff
@@ -61,9 +61,9 @@ private StrategyExecutor(OrderManager orderManager, StrategyExecutorConfiguratio
         this.maxBuysAtOnce = config.getMaxBuysAtOnce();
         this.buyStrategies = scanStrategies(BuyStrategy.class);
         this.sellStrategies = scanStrategies(SellStrategy.class);
-        sellExecutor = Executors.newFixedThreadPool(10);
+        this.sellExecutor = Executors.newFixedThreadPool(10);
         config.getExistingOrderIds().stream().forEach(orderId -> {
```
Comment (ogzd) [inline coinmx/src/main/java/com/coinmx/strategy/StrategyExecutor.java]: nit: existingBuyOrderIds to make it more clear

File: coinmx/src/main/java/com/coinmx/strategy/StrategyExecutor.java [hunk 7]
```diff
@@ -1,44 +1,71 @@
...
+    private final int maxBuysAtOnce;
+    private final AtomicInteger numCurrentOpenBuys = new AtomicInteger(0);
 
     private StrategyExecutor(OrderManager orderManager, StrategyExecutorConfiguration config) throws IOException {
         this.orderManager = orderManager;
         this.currency = Currency.getInstance(config.getCurrencyName());
+        this.maxBuysAtOnce = config.getMaxBuysAtOnce();
         this.buyStrategies = scanStrategies(BuyStrategy.class);
         this.sellStrategies = scanStrategies(SellStrategy.class);
+        this.sellExecutor = Executors.newFixedThreadPool(10);
```
Comment (ogzd) [inline coinmx/src/main/java/com/coinmx/strategy/StrategyExecutor.java]: nit: the pool size should be equal to the maxBuyAtOnce size, right?

**Pull request merged**